sympy/functions/special/holonomic.py

```diff
@@ -0,0 +1,429 @@
+    """ A function to create a Differential Operator Algebra.
+    The first arguments need to be the base polynomial ring for the algebra
+    and the second argument must be a generator.
+    Returns the class representing the operator algebra and
+    the standard operator for differentiation i.e. the `Dx` operator.
+
+    Examples
+    =======
+    from sympy.polys.domains import ZZ
+    R, Dx = DiffOperatorAlgebra
+    """
+
+    ring = DifferentialOperatorAlgebra(base_ring, generator)
+    return (ring, ring.diff_operator)
+
+
+class DifferentialOperatorAlgebra:
+    """ The class representing Differential Operator Algebra.
+    Defined by the base polynomial ring and the generator.
+
+    The attribute diff_operator is the operator `Dx` which when acted
+    on a function does differentiation.
+    """
+
+    def __init__(self, base_ring, generator):
+
+        self.base_ring = base_ring
+        self.diff_operator = DifferentialOperator([0, 1], self, generator)
+        self.gen_str = generator
+
+    def __str__(self):
+
+        string = 'Univariate Differential Operator Algebra in intermediate '\
+            + self.gen_str + ' over the base ring ' + \
+            (self.base_ring).__str__()
+
+        return string
+
+    __repr__ = __str__
+
+
+class DifferentialOperator(Expr):
+    """
+    The operator can be created by providing a list of polynomials for each power of Dx
+    and the parent ring which must be an instance of DifferentialOperatorAlgebra.
+
+    An Operator can also be created easily using the operator `Dx`. See examples below.
+
+    Examples
+    ========
+
+    >>> from sympy.functions.special.holonomic import DifferentialOperator, DiffOperatorAlgebra
+    >>> from sympy.polys.domains import ZZ, QQ
+    >>> from sympy import symbols
+    >>> x = symbols('x')
+    >>> R, Dx = DiffOperatorAlgebra(ZZ.old_poly_ring(x),'Dx')
+
+    >>> DifferentialOperator([0,1,x**2], R)
+    (0) + (1)Dx + (x**2)Dx**2
+
+    One can use the `Dx` and represent differential operators more conveniently this way also.
+    >>> (x*Dx*x + 1 - Dx**2)**2
+    (2*x**2 + 2*x + 1) + (4*x**3 + 2*x**2 - 4)Dx + (x**4 - 6*x - 2)Dx**2 + (-2*x**2)Dx**3 + (1)Dx**4
+
+    """
+
+    _op_priority = 30
+
+    def __init__(self, list_of_poly, parent_ring, generator=None, var=None):
+
+        if generator is None:
+            self.gen_symbol = symbols('Dx', commutative=False)
+        else:
+            if isinstance(generator, str):
+                self.gen_symbol = symbols(generator, commutative=False)
+            elif isinstance(generator, Symbol):
+                self.gen_symbol = generator
+
+        self.parent_ring = parent_ring
+
+        if isinstance(list_of_poly, list):
+            for i, j in enumerate(list_of_poly):
+                if isinstance(j, int):
+                    list_of_poly[i] = (
+                        (self.parent_ring).base_ring).from_sympy(S(j))
+                elif not isinstance(j, DMP):
```

Comment:
Instead of ``DMP`` we could have ``parent_ring.base_ring.dtype``. That would be more flexible. (BTW, maybe it could be possible to drop the ``_ring`` parts for brevity, so just ``parent.base.dtype``.)